### DISCUSSION TITLE
Make HP drain rate scale (in "real time") with current accuracy

### DISCUSSION THREAD

**Discussion by Yamakyu** [ppy/osu#13675] [Ideas]
Greetings

↓ SHORT VERSION ↓

The idea is to change HP loss so that players die less in beatmaps they're good at (= high accuracy while playing) and more in beatmaps they're bad at/not skilled enough for (= low accuracy while playing).

------------------------------------------

↓ LONG VERSION ↓

Among things that were changed in Lazer, I used to have one major gripe with osu! classic and it was the main reason I quit : the HP drain rate. It's not that it was too strong, but it behaved in a way that you could 

1. die on a 100/good
2. die on slider
3. die in a silence (between two circles)
4. die because of 1 difficult part in a beatmap (or 2 consecutive moderately difficult parts)

It feels like points 2 and 3 have been changed in Lazer, which is amazing and makes the game much more fun. Point 1 and 4 are still true, however the change in drain rate behavior made it also easier in a lot of beatmaps to simply not die (we'll come back to this shortly), which makes these less punishing. I don't know the exact changes inducing this, but I suspect it has to do with how much HP you loose on consecutive (or almost) misses.

I would like to suggest an additional change to HP drain rate, to make it scale according to your current accuracy % in a beatmap. 

**I currently have 2 major issues with Lazer HP drain rate.** Both seem to be tied to beatmap length.

1. Because HP drain rate is overall less punishing than in classic, longer beatmaps (which seem to have a naturally less punishing drain rate than short maps) are overall a easier. And I think it's good. It helps a lot with making you able to beat a beatmap with one very difficult part that would kill you otherwise (point 4 above). My gripe with this is it makes it possible for a player to **beat a beatmap they simply don't have the skill to beat**. Simply put, _I do not believe I have any business clearing a beatmap on which I can't at least reach 75% accuracy or so_ (for example : [this one](https://osu.ppy.sh/beatmapsets/79498#osu/221777) ([pic](https://i.imgur.com/NW04YDw.png)), [this one](https://osu.ppy.sh/beatmapsets/796330#osu/1829146) ([pic](https://i.imgur.com/LNlsYxj.png)) and [this one](https://osu.ppy.sh/beatmapsets/412938#osu/895643) ([pic](https://i.imgur.com/9ny8icW.png)), also a lot more). 75% accuracy is a lot of errors, and I believe if I make that much I'm not ready to clear the said beatmap (for the record, in classic I die in less than 30 secs of all of these), yet Lazer let me because in its generousness it's... well, _too generous_. There's already mods for running beatmaps too hard for you. 
2. HP drain rate still _seems_ to scale with beatmap length (I actually have no evidence of this, besides the fact that short beatmaps are vastly more punishing than longer ones, at equal HP drain rate stat), and in Lazer it makes some beatmaps borderline unbeatable if you don't (almost) full combo them, when they can be cleared fine in classic (i.e. the exact opposite on my previous point). Lazer makes it so that unless you almost full combo the thing, you will continuously loose HP, even if you have great accuracy for an extended period of time. I beat [this beatmap](https://osu.ppy.sh/beatmapsets/346872#osu/765567) and [this one](https://osu.ppy.sh/beatmapsets/228789#osu/778732) at respectively 94.40% and 93% accuracy (which I think is good) in classic, but cannot _clear_ them in Lazer. In fact, while I do not have screenshots as evidence, I recall dying to both of these in Lazer while having more than 96% accuracy, because my HP barely ever goes up (except in streams). For the later one I also beat [the next diff](https://osu.ppy.sh/beatmapsets/228789#osu/532666) in Lazer (not in classic), despite it being 0.6* harder than one I linked. And then again, there are already mods for making beatmaps harder for you.

My point here is that while it's fine in a lot of beatmaps (and actually better balanced I think), for some others Lazer HP drain rate reaches extremes where I believe it's either too punishing (to the point where HP lost is lost practically forever unless you full combo all the way) or too forgiving (to the point where you beat stuff you're simply not skilled enough to beat).

This is how I see it.

**The solution I propose is this.**
Change the HP drain rate behavior in Lazer so that you loose health **on misses** or 100/50 hits (natural drain rate shouldn't be affected by this change) depending on how well you're doing so far. In other words, I propose a skilled player playing a beatmap they're good at should be less punished for mistakes, and a player attempting a beatmap too hard for them should be punished """"""more"""""". How much exactly would be beyond me as I do not have the data to propose appropriate number, nor the knowledge on how to get that data. 
But to illustrate the idea, here's a proposal for scaling

- Accuracy of 95% or more → a miss/50/100 should remove a bunch less HP than it currently does (or restore more). Maybe 25% less. 
- Accuracy between 92% and 94.99% → a miss/50/100 should remove _a bit less_ HP than it currently does (or restore more). Maybe 10% less.
- Accuracy between 87% and 91.99% → HP loss/win remains unchanged.
- Accuracy between 83% and 86.99% → a miss/50/100 should remove a bit more HP than it currently does (or restore less). Maybe 5% more.
- Accuracy between 75% and 82.99% → a miss/50/100 should remove more HP than it currently does (or restore less). Maybe 10% more.
- Accuracy below 74.99% → a miss/50/100 should remove a tad more HP than it currently does (or restore less). Maybe 15%

Of course, in beatmaps in which HP drain rates is such that a 50/100 doesn't actually remove HP, it should change the HP restored accordingly → good accuracy rewards more HP upon clicking a circle, bad accuracy rewards less.
Because accuracy is an average you also cannot apply this change for the very first circles of a beatmap, because you're very likely to have accuracy of 75% or less if you only clicked like less than 10-15 circles or so, making a lot of beatmaps extremely difficult at the very beginning.

The idea here is to make it so short beatmaps with naturally (??) more punishing HP loss aren't harder to beat if you're good at them. Or in other words, I don't think you deserve to die in a beatmap if your accuracy is about 94% and you're actively making good clicks (of course if you stop playing, you die). I think 94% shows very fine skill still, and doesn't necessarily deserve punishment. On the other hand, difficult beatmaps for you will remain difficult, and will be actually difficult to beat, since Lazer won't prevent you to die (unlike what it currently does).
This also address my gripe with classic osu! in which even if you're really good at given beatmap (i.e. high accuracy) it won't necessarily kill you because it has 1 (one) short section that is very difficult ; but if you're not skilled enough for the said beatmap as a whole (i.e. low accuracy), a difficult section will end you.

This might not the way everyone sees it. This is how I see it and how I believe it should be.

I understand if this is out of scope or if it doesn't match the vision for Lazer, or if it's too late for this kind of design change.

Take care ~

**Comment by localcake** [comment]
Yes I agree with this

**Reply by frenzibyte** [reply]
Please refrain from posting +1 comments as that only brings noise. React to the discussion if you agree, either by the GitHub upvote button next to the discussion, or the regular reactions.

**Comment by iSimpFor** [comment]
instead of accuracy, maybe we can sample the accuracy of the previous x objects hit to penalize or boost hp drain so that people get a punishment for trying to play a map with a single large diff spike that is otherwise super easy and people who can play a map that is consistently hard will survive

also, i don't exactly agree with the accruacy thresholds presented 

i would propose these values(the effects do not penalize great/300 ratings):
95%+ on the previous x objects = 20% benefit to hp functions
90-95% on the previous x objects = 10% benefit to hp functions
85-90% on the previous x objects = no change to hp functions
80-85% on the previous x objects = 15%(or 20%) penalty to hp functions
under 80% on the previous x objects = 30%(or 40%) penalty to hp functions

personally, i think a C/D grade performance is pretty bad and should be punished

i also have an idea to have hp drain slow down a bit the less hp you have from 100% at full to 80% at empty(hp gain/penalties remain the same) to make it easier to perform your way out of being close to death
